### PR TITLE
Moderate refactoring and test revisions

### LINKED ISSUE #22: AnniversaryCalendar continued exposure
@MatthewCaseres shouldn't this have only three records with a final exposure with a `to=Date(2021,12,10)` since the study ends then and the decrement occurs after that point in time?

https://github.com/JuliaActuary/ExperienceAnalysis.jl/blob/65cb057bc8f1c72b34eaa4041339e296cefbecfb/test/anniversary_calendar.jl#L24-L38

### PR DESCRIPTION
**Pull request by alecloudenback**
Closes #22 

## API/Interface Changes:
The changes will likely not impact user code in typical usage, but this PR will bump the major version just in case.

- Positional arguments are required, keyword args all now have defaults.
  - All date-related arguments are now required to be Dates (some were previously allowed to be `nothing`) 
  - Previously, the `to` positional argument could be `nothing` and the `study_end` kwarg was required to be specified by user. 
  - `to` is now required and up to the user to determine the last date observed which is independent from the `study_end` (in the code and in theory). 
  - `study_end` and `study_start` now have defaults (`typemax(Date)` and `typemin(Date)`) respectively.

- All exposure types now return a three-element tuple. 
  - Previously, `Anniversary` and `AnniversaryCalendar` would return a third element with the field name of `policy_timestep`. For consistency, `Calendar` will as well, but it will be `missing`

- `Calendar` and `Anniversary` are now constructors for `AnniversaryCalendar` where the respective argument is `nothing`

### README/documentation

- Outline principles for determining exposures on different basis
- Note treatment of leap years

### Tests

- Comb through tests align expectation with documented principles
- Add a simulated set of experience where the goal is to recover a ~100% A/E for simulated claims if the exposures are calculated correctly (since the same assumption will be used for the simulated claims as the expected # of claims)
  - Accuracy improves further when tested with 10m census records (takes ~1.5 minutes on M3 Mac to simulate and do studies on 3 bases), but limited to 1m due to CI memory/speed limits
  - ` ExperienceAnalysis.AnniversaryCalendar{Year, Nothing}(Year(1), nothing)` seems to have a very slight (<0.5%) bias towards low A/E)? It gets better with 10m records vs 1m. Is this an expected bias for what's been tested or is it an issue? TBD

### CODE REVIEW

**Comment by alecloudenback** [comment]
@MatthewCaseres are you able to share your actexps comparison script and/or run it against this PR? 

The first post describes the changes, but for more context on the moderate refactor: By combining all the logic into the `AnniversaryCalendar` version and simply typemax/typemin the omitted value then there's only one overall flow. It did get a bit more complicated to handle the edge cases but let me know if you look at it and anything's not clear.

As for the tests, I went through and did make some changes to tests where I felt like the prior definitions weren't consistent with the principles now defined in the README. I'd start with those principles if you want to review this, and then let me know if you disagree with any of the test target changes. 

My plan is to merge and tag a new major version ~Thursday unless you raise an issue or explicitly would like some more time to review.

File: README.md [hunk 1]
```diff
@@ -219,5 +219,68 @@ exposure(
...
+  - If you take a `Calendar(Year(1))`/`Anniversary(Year(1))` exposure basis and split it into two pieces split by Anniversary / Calendar breakpoints, you need to take into account that in the latter pieces of exposure the expected claims needs to be reduced by the surviving exposures from the prior interval.
+    - This is saying that if you were to divide the last interval in `e3` into two parts, split by the anniversary date, that the second part of that exposure needs to take into account that not all lives in force on `2012-01-01` would survive past the anniversary that splits the interval. Pretend we actually know that the decrement should be `0.01` per day. Then the expected number of claims over the `(from = Date("2012-01-01"), to = Date("2012-12-31"), policy_timestep = missing)` exposure is `1 - 0.99^366 = 0.97474`. If we split the interval and did not take into account the reduced lives entering in the second part of the split exposure, then we would have `1- 0.99 ^191 + 1 - 0.99^175 = 1.6811` expected claims. To correct for this, the second term needs to be adjusted for the amount suriving from the first.
```
Comment (MatthewCaseres) [inline README.md:230]: typ0: suriving

File: README.md [hunk 2]
```diff
@@ -219,5 +219,68 @@ exposure(
...
+    - It is for this reason that ExperienceAnalyis.jl does not currently provide a way to "split" a `Calendar`/`Anniversary` exposure basis.
```
Comment (MatthewCaseres) [inline README.md:231]: typo: Analyis

File: README.md [hunk 3]
```diff
@@ -219,5 +219,68 @@ exposure(
...
+
+julia> e3 = exposure(ExperienceAnalysis.Calendar(Year(1)),Date(2011,07,10),Date(2012,06,15),true)
+2-element Vector{@NamedTuple{from::Date, to::Date, policy_timestep::Missing}}:
+ (from = Date("2011-07-10"), to = Date("2011-12-31"), policy_timestep = missing)
+ (from = Date("2012-01-01"), to = Date("2012-12-31"), policy_timestep = missing)
+ ```
+
+## Leap Years
+
+When a policy is issued on a leap day (February 29th), it is preferable to have the next policy year start on the 28th. This is as opposed to having the segment begin on March 1st becuase when the leap year does come around again, we wouldn't want the segment to end on February 29th.
```
Comment (MatthewCaseres) [inline README.md:253]: typo: becuase

File: README.md [hunk 4]
```diff
@@ -219,5 +219,68 @@ exposure(
...
+ (from = Date("2012-01-01"), to = Date("2012-12-31"), policy_timestep = missing)
+ ```
+
+## Leap Years
+
+When a policy is issued on a leap day (February 29th), it is preferable to have the next policy year start on the 28th. This is as opposed to having the segment begin on March 1st becuase when the leap year does come around again, we wouldn't want the segment to end on February 29th. 
+
+### Example
+
+Exposures are caulcated like this:
```
Comment (MatthewCaseres) [inline README.md:257]: typo: caulcated

File: src/ExperienceAnalysis.jl [hunk 5]
```diff
@@ -92,154 +52,147 @@ end
...
-        right_partials::Bool = true,
-    )::Vector{NamedTuple{(:from, :to, :policy_timestep),Tuple{Date,Date,Int}}}
+        study_start=typemin(from),
+        study_end=typemax(from),
+        left_partials::Bool=true,
+        right_partials::Bool=true,
+    )::Vector{NamedTuple{(:from, :to, :policy_timestep),Tuple{Date,Date,Union{Int,Nothing}}}}
 
-Return an array of name tuples `(from=Date,to=Date,policy_timestep=Int)` of the exposure periods for the given `ExposurePeriod`s. 
+Calcualte the exposure periods and returns an array of named tuples with fields:
```
Comment (MatthewCaseres) [inline src/ExperienceAnalysis.jl:63]: typo: Calcualte

File: src/ExperienceAnalysis.jl [hunk 6]
```diff
@@ -92,154 +52,147 @@ end
...
+
+    # always start from `from` to simplify the logic, and deal with
+    # study_start and left partials at the end
+    policy_timestep = 1
+    next_pol_per, policy_timestep = if isnothing(p.pol_period)
+        typemax(from), nothing
+    else
+        pp = from + p.pol_period - Day(1)
+
+        # is_leap_day = month(from) == 2 && day(from) == 29
```
Comment (MatthewCaseres) [inline src/ExperienceAnalysis.jl:106]: are these comments supposed to still be here?

**Comment by MatthewCaseres** [comment]
The comparison script is really just computing the same data set `census_dat` from benchmarks using both packages, and scrolling through `r left join julia` and `julia left join R`. It was a lot more involved when I was diving into specific issues to raise issues, that work isn't applicable here and I recommend doing something simple.

Simpler codebase is nice. Expecting some performance regression due to 

```jl
        while first(result).to < study_start
            popfirst!(result)
        end
```
In situations where `from << to` and we are doing some monthly study being able to avoid the extra work likely reduces computation time by a significant factor.

File: src/ExperienceAnalysis.jl (hunk 5 above)
Comment (alecloudenback) [reply src/ExperienceAnalysis.jl:63]: thanks - somehow I disabled my spell checker...

File: src/ExperienceAnalysis.jl (hunk 6 above)
Comment (alecloudenback) [reply src/ExperienceAnalysis.jl]: removed
